Refactoring: Making FontCascadeFonts "free" of FontSelector

vitorroriz opened the pull request:
#### 9ab41908f1d8e47dc4ec29e7af03be97c6e8d110
<pre>
Refactoring: Making FontCascadeFonts &quot;free&quot; of FontSelector
<a href="https://bugs.webkit.org/show_bug.cgi?id=289265">https://bugs.webkit.org/show_bug.cgi?id=289265</a>
<a href="https://rdar.apple.com/146405854">rdar://146405854</a>

Reviewed by Matthieu Dubet.

FontCascadeFonts is currently store in FontCascadeCache.
FontCascadeFonts currently holds a RefPtr to its related FontSelector.

In order to make it possible to share FontCascadeFont across documents (for specific simple cases),
we need to make FontCascadeFont free of context.

This patch proposes that FontCascade should hold the FontSelector instead.
FontCascade is already responsible for interfacing between FontCascadeDescription
and FontCascadeFonts. Also, FontCascadeFonts is usually accessed through
FontCascade via functions like ::glyphDataForCharacter, ::width and others, so it
make it simpler for passing the FontSelector to FontCascadeFonts when needed.

This patch is just a refactoring preparation and should produce no changes.

* Source/WebCore/platform/graphics/FontCache.cpp:
(WebCore::FontCache::updateFontCascade):
* Source/WebCore/platform/graphics/FontCache.h:
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::FontCascade):
(WebCore::FontCascade::operator=):
(WebCore::FontCascade::operator== const):
(WebCore::FontCascade::isCurrent const):
(WebCore::FontCascade::fontSelectorVersion const):
(WebCore::FontCascade::update const):
(WebCore::FontCascade::glyphDataForCharacter const):
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::primaryFont const):
(WebCore::FontCascade::fallbackRangesAt const):
(WebCore::FontCascade::isFixedPitch const):
(WebCore::FontCascade::canTakeFixedPitchFastContentMeasuring const):
(WebCore::FontCascade::fontSelector const):
* Source/WebCore/platform/graphics/FontCascadeCache.cpp:
(WebCore::FontCascadeCache::retrieveOrAddCachedFonts):
* Source/WebCore/platform/graphics/FontCascadeCache.h:
* Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
(WebCore::FontCascadeFonts::FontCascadeFonts):
(WebCore::FontCascadeFonts::determinePitch):
(WebCore::FontCascadeFonts::determineCanTakeFixedPitchFastContentMeasuring):
(WebCore::FontCascadeFonts::realizeFallbackRangesAt):
(WebCore::findBestFallbackFont):
(WebCore::FontCascadeFonts::glyphDataForSystemFallback):
(WebCore::FontCascadeFonts::glyphDataForVariant):
(WebCore::FontCascadeFonts::glyphDataForCharacter):
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/FontCascadeFonts.h:
(WebCore::FontCascadeFonts::create):
(WebCore::FontCascadeFonts::isFixedPitch):
(WebCore::FontCascadeFonts::canTakeFixedPitchFastContentMeasuring):
(WebCore::FontCascadeFonts::primaryFont):
(WebCore::FontCascadeFonts::fontSelector): Deleted.
(WebCore::FontCascadeFonts::fontSelector const): Deleted.
(WebCore::FontCascadeFonts::fontSelectorVersion const): Deleted.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::fontCascade):
* Source/WebCore/rendering/style/RenderStyle.h:

Canonical link: <a href="https://commits.webkit.org/291745@main">https://commits.webkit.org/291745@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3615085e2f3146a4fa5b3df06ef3d23c95c1faeb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93847 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/13430 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3162 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98853 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/44373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95897 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13725 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21863 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/71621 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/44373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96849 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10195 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/84798 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51958 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9879 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2433 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43688 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80144 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2509 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100889 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20899 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21151 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80754 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79985 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19915 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24530 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/1896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/14056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20883 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/26061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20570 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24030 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22311 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -1929,8 +1937,8 @@ TextStream& operator<<(TextStream& ts, const FontCascade& fontCascade)
-        ts << ", font selector " << fontCascade.fonts()->fontSelector();
-        ts << ", font selector version " << fontCascade.fonts()->fontSelectorVersion();
+        ts << ", font selector " << fontCascade.fontSelector();
+        ts << ", font selector version " << fontCascade.fontSelectorVersion();
```
Shoud we move those out of the `if` ? They don't depend on it anymore

vitorroriz replied:
yes, I think we should

vitorroriz replied:
Also, operator<< for FontSelector already streams version, so we don't need to explicitly stream fontCascade.fontSelectorVersion

vitorroriz committed:
Refactoring: Making FontCascadeFonts "free" of FontSelector
https://bugs.webkit.org/show_bug.cgi?id=289265
rdar://146405854

Reviewed by Matthieu Dubet.

FontCascadeFonts is currently store in FontCascadeCache.
FontCascadeFonts currently holds a RefPtr to its related FontSelector.

In order to make it possible to share FontCascadeFont across documents (for specific simple cases),
we need to make FontCascadeFont free of context.

This patch proposes that FontCascade should hold the FontSelector instead.
FontCascade is already responsible for interfacing between FontCascadeDescription
and FontCascadeFonts. Also, FontCascadeFonts is usually accessed through
FontCascade via functions like ::glyphDataForCharacter, ::width and others, so it
make it simpler for passing the FontSelector to FontCascadeFonts when needed.

This patch is just a refactoring preparation and should produce no changes.

* Source/WebCore/platform/graphics/FontCache.cpp:
(WebCore::FontCache::updateFontCascade):
* Source/WebCore/platform/graphics/FontCache.h:
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::FontCascade):
(WebCore::FontCascade::operator=):
(WebCore::FontCascade::operator== const):
(WebCore::FontCascade::isCurrent const):
(WebCore::FontCascade::fontSelectorVersion const):
(WebCore::FontCascade::update const):
(WebCore::FontCascade::glyphDataForCharacter const):
(WebCore::operator<<):
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::primaryFont const):
(WebCore::FontCascade::fallbackRangesAt const):
(WebCore::FontCascade::isFixedPitch const):
(WebCore::FontCascade::canTakeFixedPitchFastContentMeasuring const):
(WebCore::FontCascade::fontSelector const):
* Source/WebCore/platform/graphics/FontCascadeCache.cpp:
(WebCore::FontCascadeCache::retrieveOrAddCachedFonts):
* Source/WebCore/platform/graphics/FontCascadeCache.h:
* Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
(WebCore::FontCascadeFonts::FontCascadeFonts):
(WebCore::FontCascadeFonts::determinePitch):
(WebCore::FontCascadeFonts::determineCanTakeFixedPitchFastContentMeasuring):
(WebCore::FontCascadeFonts::realizeFallbackRangesAt):
(WebCore::findBestFallbackFont):
(WebCore::FontCascadeFonts::glyphDataForSystemFallback):
(WebCore::FontCascadeFonts::glyphDataForVariant):
(WebCore::FontCascadeFonts::glyphDataForCharacter):
(WebCore::operator<<):
* Source/WebCore/platform/graphics/FontCascadeFonts.h:
(WebCore::FontCascadeFonts::create):
(WebCore::FontCascadeFonts::isFixedPitch):
(WebCore::FontCascadeFonts::canTakeFixedPitchFastContentMeasuring):
(WebCore::FontCascadeFonts::primaryFont):
(WebCore::FontCascadeFonts::fontSelector): Deleted.
(WebCore::FontCascadeFonts::fontSelector const): Deleted.
(WebCore::FontCascadeFonts::fontSelectorVersion const): Deleted.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::fontCascade):
* Source/WebCore/rendering/style/RenderStyle.h:

Canonical link: https://commits.webkit.org/291745@main
